src/components/TimePicker/TimePicker.js

```diff
@@ -114,27 +114,41 @@ function TimePicker({forwardedRef, defaultValue, onSubmit, onInputChange}) {
         [hours, minute, amPmValue, defaultValue],
     );
 
+    const resetHours = () => {
+        setHours('00');
+        setSelectionHour({start: 0, end: 0});
+    };
+
+    const resetMinutes = () => {
+        setMinute('00');
+        setSelectionMinute({start: 0, end: 0});
+    };
+
     // This function receive value from hour input and validate it
     // The valid format is HH(from 00 to 12). If the user input 9, it will be 09. If user try to change 09 to 19 it would skip the first character
     const handleHourChange = (text) => {
```

Comment:
The logic here is a little hard to follow since the if conditions are not very alike. I have simplified the logic a little bit for more readability let me know if that would work:

```js
    const handleHourChange = (text) => {
        const trimmedText = text.trim();

        if (_.isEmpty(trimmedText)) {
            resetHours();
            return;
        }

        const isOnlyNumericValue = /^\d+$/.test(trimmedText);
        if (!isOnlyNumericValue) {
            return;
        }

        let newHour = hours;
        let newSelection = selectionHour.start;

        // When the user is inserting a digit the text would consist of three numbers
        const isInsertMode = trimmedText.length === 3;
        const isReplaceMode = trimmedText.length === 2;
        const isDeleteMode = trimmedText.length === 1;

        if (isInsertMode && selectionHour.start === 0) {
            newHour = `${trimmedText[0]}${trimmedText[2] || 0}`;
            newSelection = 1;
        } else if (isInsertMode && selectionHour.start === 1) {
            newHour = `${trimmedText[0]}${trimmedText[1] || 0}`;
            newSelection = 2;
        } else if (isReplaceMode) {
            newHour = trimmedText;
            newSelection = 2;
        } else if (isDeleteMode && selectionHour.end === 1) {
            newHour = `0${trimmedText}`;
            newSelection = 0;
        } else if (isDeleteMode && selectionHour.end === 2) {
            newHour = `${trimmedText}0`;
            newSelection = 1;
        }

        if (newHour > 24) {
            newHour = hours;
            newSelection = selectionHour.start;
        } else if (newHour > 12) {
            newHour = String(newHour - 12).padStart(2, '0');
            newSelection = 2;
            setAmPmValue(CONST.TIME_PERIOD.PM);
        }

        setHours(newHour);
        setSelectionHour({start: newSelection, end: newSelection});
        if (newSelection === 2) {
            focusMinuteInputOnFirstCharacter();
        }
    };
```

Let's do the same with the minutes case

Reply:
This looks more clear, but the conditions were designed in this way specifically to cover as many edge cases as possible.

For example, with this simplified approach we cannot replace hours with a value like 4:

https://github.com/Expensify/App/assets/12595293/09ff4f15-5303-4b75-a948-2db2c29a99ef

It's now missing the branch of checking the value to be in the range of  0..1, and 2..12.

Also, this naming is a bit misleading, because the length will be 1 not only on delete, but on replacing the selection with a single digit:

```js
const isDeleteMode = trimmedText.length === 1;
```